bpf/estat/metaslab-alloc.c

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright 2020 Delphix. All rights reserved.
+ *
+ * SPDX-License-Identifier: GPL-2.0-or-later
+ */
+
+#include <sys/metaslab.h>
+#include <sys/vdev_impl.h>
+#include <sys/spa_impl.h>
+
+#define VD_NAME_SIZE 32
+typedef struct {
+	u64 ts;
+	u64 size;
+	u64 asize;
+	u64 alloc_time;
+	char vd_name[VD_NAME_SIZE];
+} data_t;
+
+BPF_HASH(data_map, u32, data_t);
+
+#ifndef OPTARG
+#define	POOL "domain0"
+#else
+#define	POOL (OPTARG)
+#endif
+
+static inline bool
+equal_to_pool(char *str)
+{
+	char comparand[sizeof (POOL)];
+	bpf_probe_read(&comparand, sizeof (comparand), str);
+	char compare[] = POOL;
+
+	for (int i = 0; i < sizeof (comparand); ++i)
+		if (compare[i] != comparand[i])
+			return (false);
+
+	return (true);
+}
+
+// @@ kprobe|metaslab_alloc_dva|metaslab_alloc_dva_entry
+int
+metaslab_alloc_dva_entry(struct pt_regs *ctx,
+	spa_t *spa, metaslab_class_t *mc, uint64_t psize)
+{
+	u32 tid = bpf_get_current_pid_tgid();
+	data_t data = {};
+
+	if (!equal_to_pool(spa->spa_name))
+		return (0);
+
+	data.ts = bpf_ktime_get_ns();
+	data.size = psize;
+
+	data_map.update(&tid, &data);
+
+	return (0);
+}
+
+// @@ kprobe|vdev_allocatable|vdev_allocatable_entry
+int
+vdev_allocatable_entry(struct pt_regs *ctx, vdev_t *vd)
+{
+	u32 tid = bpf_get_current_pid_tgid();
+	data_t *data = data_map.lookup(&tid);
+
+	if (data == NULL || data->ts == 0)
+		return (0);
+
+	if (vd->vdev_path != NULL) {
+		bpf_probe_read_str(data->vd_name, VD_NAME_SIZE, vd->vdev_path);
+	} else {
+		bpf_probe_read_str(data->vd_name, VD_NAME_SIZE, vd->vdev_ops->vdev_op_type);
+	}
+
+	return (0);
+}
+
+// @@ kprobe|metaslab_group_alloc|metaslab_group_alloc_entry
+int
+metaslab_group_alloc_entry(struct pt_regs *ctx,
+	metaslab_group_t *mg, zio_alloc_list_t *zal, uint64_t asize)
+{
+	u32 tid = bpf_get_current_pid_tgid();
+	data_t *data = data_map.lookup(&tid);
+
+	if (data == NULL || data->ts == 0)
+		return (0);
+
+	data->asize = asize;
+	data->alloc_time = bpf_ktime_get_ns();
+
+	data_map.update(&tid, data);
```

Comment:
why do we need to call .update() here, but not in vdev_allocatable_entry()?

Reply:
Good question, I'm not sure.. @brad-lewis can you comment on when we need to call `.update`, and when it's not necessary? do you know?

Reply:
So map.lookup() gives you a pointer to the value in the map if it exists.   So then you can update that value in place without a call to map.update().  So it's not needed here and should be removed.  

The update call overwrites the value in the map.   So if you don't care if there is a value, you can use update to initialize/override.

Reply:
So, based on that, I should remove `update` from the `metaslab_group_alloc_entry` probe, too?

Reply:
Sorry I wasn't clear.  The update() in metaslab_group_alloc_entry() should be removed.  The data_t struct is updated in place.   

The update() in metaslab_alloc_dva_entry() is needed as it's provided a new value to the map (overwriting in some cases) not modifying an existing value.